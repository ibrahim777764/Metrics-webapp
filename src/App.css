body {
  background-color: #f15793;
}

.App {
  text-align: center;
}
