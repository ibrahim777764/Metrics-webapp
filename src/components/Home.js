import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { FaRegArrowAltCircleRight } from 'react-icons/fa';
import { NavLink } from 'react-router-dom';
import CountryPreview from './CountryPreview';
import ErrorBoundary from './ErrorBoundary';
import { endSearch } from '../redux/search/search';

const Home = ({ theme, countries }) => {
  const dispatch = useDispatch();
  const search = useSelector((state) => state.search, shallowEqual);

  const { hsl } = theme;
  const { searchOpen, searchResult } = search;

  const handleClose = () => {
    dispatch(endSearch());
  };

  return (
    <main className="h-full w-full relative top-16 z-0 grid grid-cols-2 sm:grid-cols-4 lg:grid-cols-6 xl:grid-cols-8">
      {
        countries.map((country) => (
          <CountryPreview
            key={country.cca2.toLowerCase()}
            hsl={hsl}
            country={country}
          />
        ))
      }
      <ErrorBoundary>
        {
          searchOpen && (searchResult.length !== 0) && (
            <section className="absolute rounded overflow-hidden shadow-2xl w-10/12 md:w-5/12 z-20 top-3 right-3 bg-white text-blue-500 flex flex-col">
              {
                searchResult.map((country) => (
                  <NavLink
                    to={`/${country.name.common.toLowerCase()}/details`}
                    className="flex justify-between even:bg-slate-100 p-2 items-center"
                    key={country.cca2}
                    onClick={handleClose}
                  >
                    <span>
                      {country.name.common}
                      {', '}
                      {country.capital[0]}
                    </span>
                    <FaRegArrowAltCircleRight className="text-base ml-1" />
                  </NavLink>
                ))
              }
            </section>
          )
        }
      </ErrorBoundary>
    </main>
  );
};

Home.propTypes = {
  theme: PropTypes.instanceOf(Object).isRequired,
  countries: PropTypes.instanceOf(Array).isRequired,
};

export default Home;
