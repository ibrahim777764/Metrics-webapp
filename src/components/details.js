import { useParams } from 'react-router-dom';

const Details = (props) => {
  const data = props;
  const country = useParams();
  return (
    <div className="details">
      <div className="contry-div">
        <h1>
          Country:
          {country.country}
        </h1>
      </div>
      <div className="info">
        <div className="details-div">
          <div className="dates">
            <h2 className="date-head">DETAILS</h2>
            <p>
              <b>Capital:</b>
              { data.country.All?.capital_city}
            </p>
          </div>
          <div className="main-wrap">
            <div className="sor">
              <p className="dat-type">TOTAL CONFIRMED:</p>
              <b className="dat-type">{data.country.All?.confirmed}</b>
            </div>
            <div className="sor">
              <p className="dat-type">TOTAL DEATH:</p>
              <b className="dat-type">{data.country.All?.deaths}</b>
            </div>
            <div className="sor">
              <p>Life_expectancy:</p>
              <b className="dat-type">{ data.country.All?.life_expectancy}</b>
            </div>
            <div className="sor">
              <p>population:</p>
              <b className="dat-type">{ data.country.All?.population}</b>
            </div>
            <div className="sor">
              <p>Latitude:</p>
              <b className="dat-type">{data.country.All?.lat }</b>
            </div>
            <div className="sor">
              <p>Longtitide:</p>
              <b className="dat-type">{ data.country.All?.long}</b>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Details;
