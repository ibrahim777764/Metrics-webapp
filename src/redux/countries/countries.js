const COUNTRIES_FETCHED = 'app/countries/COUNTRIES_FETCHED';

const fetchCountries = (countries) => ({ type: COUNTRIES_FETCHED, payload: countries });

const reducer = (state = [], action) => {
  const { type, payload } = action;

  switch (type) {
    case COUNTRIES_FETCHED:
      return [...payload];
    default:
      return state;
  }
};

const getCountries = () => async (dispatch) => {
  await fetch('https://restcountries.com/v3.1/all')
    .then((data) => data.json())
    .then((data) => {
      const countries = [];
      data.forEach((country) => {
        const {
          name,
          cca2,
          capital,
        } = country;

        const formattedCountry = {
          name,
          cca2,
          capital,
        };

        countries.push(formattedCountry);
      });
      dispatch(fetchCountries(countries));
    });
};

export { getCountries };
export default reducer;
