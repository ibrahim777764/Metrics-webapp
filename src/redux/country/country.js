const COUNTRY_SELECTED = 'app/country/COUNTRY_SELECTED';

const selectCountry = (payload) => ({ type: COUNTRY_SELECTED, payload });

const reducer = (state = {}, action) => {
  const { type, payload } = action;

  switch (type) {
    case COUNTRY_SELECTED:
      return {
        ...payload,
      };
    default:
      return state;
  }
};

const getCountryData = (iso) => async (dispatch) => {
  await fetch(`https://restcountries.com/v3.1/alpha/${iso}`)
    .then((data) => data.json())
    .then((data) => {
      const {
        name,
        cca2,
        cca3,
        independent,
        unMember,
        currencies,
        capital,
        region,
        subregion,
        languages,
        latlng: coordinates,
        landlocked,
        area,
        population,
        fifa,
        car: { side: drivingSide },
        timezones,
        continents,
        flags: { svg: flag },
        coatOfArms: { svg: emblem },
        startOfWeek,
      } = data[0];

      const formattedData = {
        name,
        cca2,
        cca3,
        independent,
        unMember,
        currencies,
        capital,
        region,
        subregion,
        languages,
        coordinates,
        landlocked,
        area,
        population,
        fifa,
        drivingSide,
        timezones,
        continents,
        flag,
        emblem,
        startOfWeek,
      };

      dispatch(selectCountry(formattedData));
    });
};

export { getCountryData };
export default reducer;
