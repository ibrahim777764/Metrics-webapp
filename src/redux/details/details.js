const DETAILS_OPENED = 'app/details/DETAILS_OPENED';
const DETAILS_CLOSED = 'app/details/DETAILS_CLOSED';

const openDetails = () => ({ type: DETAILS_OPENED, payload: true });
const closeDetails = () => ({ type: DETAILS_CLOSED, payload: false });

const reducer = (state = false, action) => {
  const { type, payload } = action;

  switch (type) {
    case DETAILS_OPENED:
      return payload;
    case DETAILS_CLOSED:
      return payload;
    default:
      return state;
  }
};

export { openDetails, closeDetails };
export default reducer;
