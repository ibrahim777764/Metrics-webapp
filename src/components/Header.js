import React from 'react';
import PropTypes from 'prop-types';
import { FaSearch, FaChevronLeft } from 'react-icons/fa';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { startSearch } from '../redux/search/search';
import { closeDetails } from '../redux/details/details';
import Form from './Form';

const Header = (props) => {
  const dispatch = useDispatch();
  const detailsOpen = useSelector((state) => state.details);
  const search = useSelector((state) => state.search, shallowEqual);
  const { theme } = props;
  const { hex } = theme;
  const style = {
    backgroundColor: hex,
  };

  const { searchOpen } = search;

  const handleClick = () => {
    dispatch(closeDetails());
  };

  const handleSearch = () => {
    dispatch(startSearch());
  };

  return (
    <header className="w-full flex fixed top-0 z-10 shadow-md items-center justify-between px-6 py-4" style={style}>
      {!detailsOpen && (
        <>
          <h1 className="text-3xl text-white font-bold">
            <span className="text-black">Cultur</span>
            escape
          </h1>
          <section className="flex items-center text-2xl">
            {!searchOpen && (
              <button type="button" onClick={handleSearch}>
                <FaSearch />
              </button>
            )}
            {searchOpen && <Form />}
          </section>
        </>
      )}
      {detailsOpen && (
        <NavLink to="/" className="flex items-center text-2xl" onClick={handleClick}>
          <FaChevronLeft className="mr-1" />
          {' '}
          Home
        </NavLink>
      )}
    </header>
  );
};

Header.propTypes = {
  theme: PropTypes.instanceOf(Object).isRequired,
};

export default Header;
