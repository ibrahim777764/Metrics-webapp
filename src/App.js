import React, { useEffect } from 'react';
import { Routes, Route } from 'react-router-dom';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import Header from './components/Header';
import Home from './components/Home';
import CountryDetails from './components/CountryDetails';
import { getCountries } from './redux/countries/countries';

const App = () => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getCountries());
  }, []);
  const theme = useSelector((state) => state.theme, shallowEqual);
  const countries = useSelector((state) => state.countries, shallowEqual);

  return (
    <div className="w-screen h-screen relative overflow-x-hidden flex flex-col text-white">
      <Header theme={theme} />
      <Routes>
        <Route
          path="/"
          element={(<Home theme={theme} countries={countries} />)}
        />
        <Route
          path="/:country/details"
          element={(<CountryDetails theme={theme} countries={countries} />)}
        />
      </Routes>
    </div>
  );
};

export default App;
