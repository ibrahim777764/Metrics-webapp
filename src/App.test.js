import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { combineReducers, createStore, applyMiddleware } from 'redux';
import { BrowserRouter as Router } from 'react-router-dom';
import thunk from 'redux-thunk';
import country from './redux/country/country';
import search from './redux/search/search';
import countries from './redux/countries/countries';
import details from './redux/details/details';
import theme from './redux/theme/theme';
import App from './App';
import Header from './components/Header';
import Home from './components/Home';

const rootReducer = combineReducers({
  countries,
  country,
  search,
  details,
  theme,
});

const Store = createStore(
  rootReducer,
  applyMiddleware(thunk),
);

describe('Header component', () => {
  test('renders app title', () => {
    const theme = {
      hex: '#ffffff',
      hsl: 'hsl(360,50,',
    };

    render(
      <Provider store={Store}>
        <Router>
          <Header theme={theme} />
        </Router>
      </Provider>,
    );
    const appTitle = screen.getByText(/escape/i);
    expect(appTitle).toBeInTheDocument();
  });
});

describe('Home page component', () => {
  let countries;
  let theme;

  beforeEach(() => {
    countries = [
      { name: { common: 'congo', official: 'dr congo' }, cca2: 'cd', capital: ['kinshasa'] },
      { name: { common: 'rwanda', official: 'rwanda' }, cca2: 'rw', capital: ['kigali'] },
      { name: { common: 'burundi', official: 'burundi' }, cca2: 'bu', capital: ['bujumbura'] },
    ];
    theme = {
      hex: '#ffffff',
      hsl: 'hsl(360,50,',
    };
  });

  test('renders correctly', () => {
    const homePage = renderer.create(
      <Provider store={Store}>
        <Router>
          <App>
            <Home theme={theme} countries={countries} />
          </App>
        </Router>
      </Provider>,
    ).toJSON();
    expect(homePage).toMatchSnapshot();
  });

  test('renders  country  name', () => {
    render(
      <Provider store={Store}>
        <Router>
          <Home theme={theme} countries={countries} />
        </Router>
      </Provider>,
    );
    const countryName = screen.getAllByText(/congo/i);
    expect(countryName).toHaveLength(1);
  });

  test('Country Preview link', () => {
    const { container } = render(
      <Provider store={Store}>
        <Router>
          <Home theme={theme} countries={countries} />
        </Router>
      </Provider>,
    );

    const countryPreview = container.querySelector('a');
    userEvent.click(countryPreview);
    expect(screen.getByText(/congo/i)).toBeInTheDocument();
  });

  test('renders  country  capital', () => {
    render(
      <Provider store={Store}>
        <Router>
          <Home theme={theme} countries={countries} />
        </Router>
      </Provider>,
    );
    const countryCapital = screen.getAllByText(/kinshasa/i);
    expect(countryCapital).toHaveLength(1);
  });
});
