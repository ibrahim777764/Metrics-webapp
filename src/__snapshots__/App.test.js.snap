// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page component renders correctly 1`] = `
<div
  className="w-screen h-screen relative overflow-x-hidden flex flex-col text-white"
>
  <header
    className="w-full flex fixed top-0 z-10 shadow-md items-center justify-between px-6 py-4"
    style={
      Object {
        "backgroundColor": "#4369b2",
      }
    }
  >
    <h1
      className="text-3xl text-white font-bold"
    >
      <span
        className="text-black"
      >
        Cultur
      </span>
      escape
    </h1>
    <section
      className="flex items-center text-2xl"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </button>
    </section>
  </header>
  <main
    className="h-full w-full relative top-16 z-0 grid grid-cols-2 sm:grid-cols-4 lg:grid-cols-6 xl:grid-cols-8"
  />
</div>
`;
